### PR TITLE
April tag 1 with pose estimator 

### PR DESCRIPTION
**Pull request by Arima25**

### CODE REVIEW

File: FaceAprilTag.txt
Comment (isaacngkh) [inline FaceAprilTag.txt:1]: Is this file needed? If not, should delete

File: src/main/java/frc/robot/subsystems/LimeVision/LimeLightSub.java [hunk 1]
```diff
@@ -52,9 +85,22 @@ public class LimeLightSub extends SubsystemBase {
   public LimeLightSub(String limelight_networktable_name) {
     limelight_comm = new LimeLightComms(limelight_networktable_name);
     limelight_comm.set_entry_double("ledMode", 3);
+    Shuffleboard.getTab("Limelight").addDouble("BotPose TX", ()->getBlueBotPose()[0]);
```
Comment (isaacngkh) [inline src/main/java/frc/robot/subsystems/LimeVision/LimeLightSub.java]: Can we remove the Shuffleboard stuff if it is just used for testing?

File: src/main/java/frc/robot/subsystems/LimeVision/LimeLightSub.java [hunk 2]
```diff
@@ -64,16 +110,44 @@ public void periodic() {
     SmartDashboard.putNumber("tx", tx.getDouble(0));
     SmartDashboard.putNumber("ty", ty.getDouble(0));
     SmartDashboard.putNumber("ta", ta.getDouble(0));
-    SmartDashboard.putNumber("theta", getTheta());
-    SmartDashboard.putNumber("AngleToTarget", getAngle());
+    // SmartDashboard.putNumber("theta", getTheta());
```
Comment (isaacngkh) [inline src/main/java/frc/robot/subsystems/LimeVision/LimeLightSub.java]: Remove comment/SmartDashboard if just for testing

File: src/main/java/frc/robot/subsystems/LimeVision/LimeLightSub.java [hunk 3]
```diff
@@ -64,16 +110,44 @@ public void periodic() {
...
+    // // displaying error values
+    // SmartDashboard.putNumber("Error Angle", getError());
+    SmartDashboard.putNumber("DistanceX Error X", getErrorFromMegaTagX());
+    SmartDashboard.putNumber("DistanceX Error Y", getErrorFromMegaTagY());
     
-    // displaying error values
-    SmartDashboard.putNumber("Error Angle", getError());
+  
+    // botpose
+    if (hasTarget())
```
Comment (isaacngkh) [inline src/main/java/frc/robot/subsystems/LimeVision/LimeLightSub.java]: Remove if not needed

File: src/main/java/frc/robot/subsystems/LimeVision/LimeLightSub.java [hunk 4]
```diff
@@ -100,6 +174,27 @@ public double getAngle() {
     return angle;
   }
 
+  public double[] getBotPose() {
```
Comment (isaacngkh) [inline src/main/java/frc/robot/subsystems/LimeVision/LimeLightSub.java]: I believe we are not using this correct? We are using the new bot pose blue instead. We should remove this to avoid confusion.

File: src/main/java/frc/robot/subsystems/LimeVision/LimeLightSub.java [hunk 5]
```diff
@@ -117,4 +212,111 @@ public boolean checkPipe() {
...
+        else{
+          return;
+        }
+          stds.set(0,0,xyStds);
+          stds.set(1,0,xyStds);
+          stds.set(2,0, degStds);
+          Rotation2d degree = new Rotation2d(temp[5]* Math.PI / 180);
+          Pose2d newPose = new Pose2d(temp[0],temp[1],degree); //creates new pose2d with limelight data
+          
+          VisionContainer.drivebase.addActualVisionReading(newPose ,Timer.getFPGATimestamp() - (temp[6]/1000.0),stds); //Changes standard dev base on apriltags and drive
```
Comment (isaacngkh) [inline src/main/java/frc/robot/subsystems/LimeVision/LimeLightSub.java]: We should comment this line of code first. It will crash the main branch when VisionContainer's drivebase is not initialized.

Should use the GameRobotContainer's drivebase instead.

File: src/main/java/frc/robot/subsystems/LimeVision/LimeLightSub.java [hunk 6]
```diff
@@ -117,4 +212,111 @@ public boolean checkPipe() {
...
+      colorBotPose[9] = blueBotPose[9];
+      colorBotPose[11] = getTID()+1;
+
+    }
+    
+
+    return colorBotPose; 
+  }
+
+  public double[] getTargetSpace(){
```
Comment (isaacngkh) [inline src/main/java/frc/robot/subsystems/LimeVision/LimeLightSub.java]: Is this method used anywhere?

File: src/main/java/frc/robot/testcontainers/VisionContainer.java [hunk 7]
```diff
@@ -30,12 +47,13 @@ public class VisionContainer implements BaseContainer
 {
 
   // The robot's subsystems and commands are defined here...
-  private final SwerveSubsystem drivebase;
+  public static SwerveSubsystem drivebase;
```
Comment (RobertTheBlair) [inline src/main/java/frc/robot/testcontainers/VisionContainer.java]: as discussed, will convert this back to a non-static, and pass as parameter to LimeLightSub, so it is easy to use with any container, not just VisionContainer

File: src/main/java/frc/robot/subsystems/LimeVision/LimeLightSub.java [hunk 8]
```diff
@@ -117,4 +212,111 @@ public boolean checkPipe() {
...
+  
+    int tid = (int)this.tid.getDouble(-1)-1;
+    
+    return tid;
+    
+  }
+
+  public void setData(){
+    double[] temp = getBlueBotPose();
+    Pose2d currentPose = VisionContainer.drivebase.getPose();
```
Comment (RobertTheBlair) [inline src/main/java/frc/robot/subsystems/LimeVision/LimeLightSub.java]: The drivebase should be an member of the LimeLightSub class, not accessed via a global variable

File: src/main/java/frc/robot/subsystems/LimeVision/LimeLightSub.java (hunk 5 above)
Comment (RobertTheBlair) [reply src/main/java/frc/robot/subsystems/LimeVision/LimeLightSub.java]: Or pass in as a parameter to this subsystem.

**Review by RobertTheBlair — CHANGES_REQUESTED** [review]
Needs changes to not use static variables.  And in the end, will need new constants for new robot

File: src/main/java/frc/robot/subsystems/LimeVision/LimeLightSub.java (hunk 1 above)
Comment (RobertTheBlair) [reply src/main/java/frc/robot/subsystems/LimeVision/LimeLightSub.java]: Adding a 'verbose' flag

**Review by RobertTheBlair — APPROVED** [review]
looks good.   Will need to update with real parameters on for the new robot

**Review by RobertTheBlair — APPROVED** [review]
Seems good
